actors/migration/nv4/test/correct_power_accounting_test.go

```diff
@@ -0,0 +1,472 @@
+			Deadline: oDlInfo.Index,
+			Partitions: []miner0.PoStPartition{{
+				Index:   oPIdx,
+				Skipped: bitfield.New(),
+			}},
+			Proofs: []proof.PoStProof{{
+				PoStProof: abi.RegisteredPoStProof_StackedDrgWindow32GiBV1,
+			}},
+			ChainCommitEpoch: oDlInfo.Challenge,
+			ChainCommitRand:  []byte("not really random"),
+		})
+
+		v2, err = v2.WithEpoch(oDlInfo.Last())
+		require.NoError(t, err)
+		vm2.ApplyOk(t, v2, builtin.SystemActorAddr, builtin.CronActorAddr, big.Zero(), builtin.MethodsCron.EpochTick, nil)
+
+		// advance to upgrade period and submit post
+		dlInfo, pIdx, v2 = vm2.AdvanceTillProvingDeadline(t, v2, minerAddrs.IDAddress, upgradeSectorNumber)
+		require.True(t, dlInfo.Last() > oDlInfo.Last())
+
+		vm2.ApplyOk(t, v2, worker, minerAddrs.RobustAddress, big.Zero(), builtin.MethodsMiner.SubmitWindowedPoSt, &miner2.SubmitWindowedPoStParams{
+			Deadline: dlInfo.Index,
+			Partitions: []miner2.PoStPartition{{
+				Index:   pIdx,
+				Skipped: bitfield.New(),
+			}},
+			Proofs: []proof.PoStProof{{
+				PoStProof: abi.RegisteredPoStProof_StackedDrgWindow32GiBV1,
+			}},
+			ChainCommitEpoch: dlInfo.Challenge,
+			ChainCommitRand:  []byte("not really random"),
+		})
+
+		v, err = v.WithEpoch(dlInfo.Last())
+		require.NoError(t, err)
+		vm2.ApplyOk(t, v2, builtin.SystemActorAddr, builtin.CronActorAddr, big.Zero(), builtin.MethodsCron.EpochTick, nil)
+	}
+
+	// miner still has power
+	assert.Equal(t, numSectors*32<<30, vm2.MinerPower(t, v2, minerAddrs.IDAddress).Raw.Uint64())
+
+	// miner state invariants hold
+	var st2 miner2.State
+	err = v2.GetState(minerAddrs.IDAddress, &st2)
+	require.NoError(t, err)
+
+	stateTree, err := v2.GetStateTree()
+	require.NoError(t, err)
+	totalBalance, err := v2.GetTotalActorBalance()
+	require.NoError(t, err)
+	msgs, err = states.CheckStateInvariants(stateTree, totalBalance, v2.GetEpoch())
+	require.NoError(t, err)
+
+	// The v2 migration will not correctly update power totals. Expect 5 invariant violations all related to power
+	assert.Equal(t, 5, len(msgs.Messages()), strings.Join(msgs.Messages(), "\n"))
+	for _, msg := range msgs.Messages() {
+		assert.True(t, strings.Contains(msg, "t04 power:"))
+	}
+
+	// Run nv4 migration to fix totals
+	nextRoot, err = nv4.MigrateStateTree(ctx, v.Store(), v2.StateRoot(), v2.GetEpoch(), nv4.Config{}) // v.Store() is not threadsafe
+	require.NoError(t, err)
+
+	v2nv4, err := vm2.NewVMAtEpoch(ctx, lookup, v.Store(), nextRoot, v2.GetEpoch()+1)
+	require.NoError(t, err)
+
+	// all state invariants hold
+	stateTree, err = v2nv4.GetStateTree()
+	require.NoError(t, err)
+	totalBalance, err = v2nv4.GetTotalActorBalance()
+	require.NoError(t, err)
+	msgs, err = states.CheckStateInvariants(stateTree, totalBalance, v2nv4.GetEpoch())
+	require.NoError(t, err)
+	assert.Equal(t, 0, len(msgs.Messages()), strings.Join(msgs.Messages(), "\n"))
+}
```

Comment:
Honestly I don't think this approach to testing this is very good. It's a good representation of the specific scenario we're in, but we could also just construct that much more directly, and could also construct more scenarios to demonstrate robustness of the fix to other things that might go wrong before it is run.

We're only touching the power actor. I recommend something closer to a unit test that either constructs the power state directly (since we got there by migrating it directly, but wrongly) and then inserts into a state tree with just the power actor for invoking the migration (and validation? but that will need corresponding miner state).